src/main/java/com/github/legoatoom/connectiblechains/chain/ChainLink.java

```diff
@@ -0,0 +1,323 @@
+        assert primary.world instanceof ServerWorld;
+        Vec3d startPos = start.getPos().add(start.getLeashOffset());
+        Vec3d endPos = end.getPos().add(end.getLeashOffset());
+
+        Vec3d tmp = endPos;
+        if (reverse) {
+            endPos = startPos;
+            startPos = tmp;
+        }
+
+        Vec3f offset = Helper.getChainOffset(startPos, endPos);
+        startPos = startPos.add(offset.getX(), 0, offset.getZ());
+        endPos = endPos.add(-offset.getX(), 0, -offset.getZ());
+
+        double distance = startPos.distanceTo(endPos);
+
+        double x = MathHelper.lerp(v, startPos.getX(), endPos.getX());
+        double y = startPos.getY() + Helper.drip2((v * distance), distance, endPos.getY() - startPos.getY());
+        double z = MathHelper.lerp(v, startPos.getZ(), endPos.getZ());
+
+        y += -ModEntityTypes.CHAIN_COLLISION.getHeight() + 2 / 16f;
+
+        ChainCollisionEntity c = new ChainCollisionEntity(primary.world, x, y, z, this);
+        if (primary.world.spawnEntity(c)) {
+            return c;
+        } else {
+            ConnectibleChains.LOGGER.warn("Tried to summon collision entity for a chain, failed to do so");
+            return null;
+        }
+    }
+
+    /**
+     * Finds all players that are in {@code world} and tracking either the primary or secondary.
+     *
+     * @param world the world to search in
+     * @return A set of all players that track the primary or secondary.
+     */
+    private Set<ServerPlayerEntity> getTrackingPlayers(World world) {
+        assert world instanceof ServerWorld;
+        Set<ServerPlayerEntity> trackingPlayers = new HashSet<>(
+                PlayerLookup.around((ServerWorld) world, primary.getBlockPos(), ChainKnotEntity.VISIBLE_RANGE));
+        trackingPlayers.addAll(
+                PlayerLookup.around((ServerWorld) world, secondary.getBlockPos(), ChainKnotEntity.VISIBLE_RANGE));
+        return trackingPlayers;
+    }
+
+    public boolean isDead() {
+        return !alive;
+    }
+
+    /**
+     * Returns the squared distance between the primary and secondary.
+     */
+    public double getSquaredDistance() {
+        return this.primary.squaredDistanceTo(secondary);
+    }
+
+    /**
+     * Two links are considered equal when the involved entities are the same, regardless of their designation
+     * and the links have the same living status.
+     */
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ChainLink link = (ChainLink) o;
+
+        boolean partnersEqual = primary.equals(link.primary) && secondary.equals(link.secondary) ||
+                primary.equals(link.secondary) && secondary.equals(link.primary);
+        return alive == link.alive && partnersEqual;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(primary, secondary, alive);
+    }
```

Comment:
I am not used to using `hashCode()` what is the actual use of this?

Reply:
`hashCode` is used by Sets and Maps  and such as a quick way to check equality and should always match `equals`. So if two objects `equals()` then they should also have the same hash code.